Full API documentation (#24)
r-scheele/rego_builder: docs.md

  Features
============
   - Translating JSON to REGO
   - CRUD operations for policies for persistence.
   - Supports existing policies for editing.
   - Supports creating new policies from scratch.

   Architecture
   ============
   The architecture of the API is based on the following principles:
   - The API is written in Python and FastAPI framework.
   - The API is written in a way that it is easy to use, extend, test and maintain.

![Screenshot 2022-06-13 at 09 55 43](https://user-images.githubusercontent.com/67229938/173343113-d51d72b4-84c8-4c3b-8555-af41e59cd2de.png)

View the Architecture in the browser [Here](https://www.figma.com/file/684S7kO4dPQbbZFZr6xZOn/Rego-builder?node-id=0%3A1) 

   Installation
   ============
- Manual installation - Available [Here](https://github.com/r-scheele/rego_builder#readme)

- Build your own image and start the container -
   ```bash
   $ Configure environment variables - the .env file
   $ docker-compose -f docker-compose.dev.yml up -d

   ```
- Pull the official image from Docker Hub - and start the container 

   ```bash
   $ docker pull rscheele3214/rego-builder:latest
   $ docker run rscheele3214/rego-builder:latest
   ```

> Usage

Reviewer: Before directly going into usage, I'll expect to be briefed on how the translation from JSON to REGO works. I have no idea of commands - what do they stand for as well?

I'll expect a brief/detailed section of what commands are - you can then list the various commands here.

After explaining what the commands are, you should give us a lead-in on how they are compiled to form a rule - command, and properties. There's no explanation of what properties are.

Author: Alright would submit another commit for review soon

         * [Handling * as the wildcard flag](#Handling-'*'-as-the-wildcard-flag)
         * [Allowing all path parameter after a path section, except one](#Handling-'*'-as-the-wildcard-flag)
         * [ Handling equality check between a particular property on the request object and a valueHandling * as the wildcard flag](#Handling-'*'-as-the-wildcard-flag)
         * [Allow access to a particular path](#allow-access-to-a-particular-path)
      2. [Input_props_in](#input_props_in)
      3. [Input_props_in_as](#input_props_in_as)
      4. [allow_full_access](#allow_full_access)
   * [API CRUD operations](#api-crud-operations)

  Introduction
============
This API is responsible for converting JSON variables to REGO which in turn is used in application policy management. The policy management application is born from the need to manage Open Policy Agent ( OPA ) policies written in REGO from a user interface for non-technical users and an API through JSON for technical users.

  Features
============
   - Translating JSON to REGO
   - CRUD operations for policies for persistence.
   - Supports existing policies for editing.
   - Supports creating new policies from scratch.

> 

Reviewer: Repetition of the second item on the list.

   Architecture
   ============
   The architecture of the API is based on the following principles:
   - The API is written in Python and FastAPI framework.
   - The API is written in a way that it is easy to use, extend, test and maintain.

![Screenshot 2022-06-13 at 09 55 43](https://user-images.githubusercontent.com/67229938/173343113-d51d72b4-84c8-4c3b-8555-af41e59cd2de.png)

View the Architecture in the browser [Here](https://www.figma.com/file/684S7kO4dPQbbZFZr6xZOn/Rego-builder?node-id=0%3A1) 

   Installation
   ============

> - Manual installation - Available [Here](https://github.com/r-scheele/rego_builder#readme)

Reviewer: Instead of directing users to another document, place the steps here as well.

> - Build your own image and start the container -

Reviewer: ```suggestion
- Build the application image and start the container:
```

>    ```bash

Reviewer: ```suggestion
   ```console
```

   $ Configure environment variables - the .env file
   $ docker-compose -f docker-compose.dev.yml up -d

   ```

> - Pull the official image from Docker Hub - and start the container 

Reviewer: ```suggestion
- Pull the official image from Docker Hub and start the container 
```

>    ```bash

Reviewer: ```suggestion
   ```console
```

   $ docker pull rscheele3214/rego-builder:latest
   $ docker run rscheele3214/rego-builder:latest
   ```

How the translation from JSON to REGO works?
============
The REGO policies are created by a set of functions that serves as the translation logic, They're called commands. Each of these commands is responsible for writing specific REGO rule to a `.rego` file. However, the JSON to REGO conversion must follow certain conditions as defined by a pydantic model for effectiveness. The following is an example of how a rule is defined:

   ```json     
   {
      "command": "input_prop_equals",
      "properties": {
         "input_property": "request_method",
         "value": "GET"
      }
   }
   ```
In this case, The command key represents the operation to be performed and the properties key represents the properties that are being used in the operation.
`input_prop_equals` is the command in example, which initiates the appropriate operation, on the properties object.
The above rule translates to an equality check between the input_property and the value. <br />
The REGO equivalent of the above rule object is: <br />
   `input.request_method == "GET"`

> 

Reviewer: You should start this section with a lead like this:

A rule is defined by two keys: command and properties. The command key holds one of the recognized commands and the properties hold another dictionary containing the `input_property` and `value`. In special cases, the `datasource_items` items are also used/added etc etc.

> ### As of now, the API supports the following commands:

Reviewer: No need for this to be a heading.

1. input_prop_equals
2. input_prop_in
3. input_prop_in_as
4. allow_full_access

Usage
============
### Detailed explanation of commands with examples: 

> 

Reviewer: Add a sentence or two explaining what this section will contain.

> ## 1. Input_props_equals

Reviewer: Preferably, don't number them.

   This command has different logic to handle series of equality checks.
  ### - Handling '*' as the wildcard flag: <br /> 

   This logic handles all the paths after a particular section. if `/collections/` is supplied as the option, all the routes after it will be allowed e.g allow `/collections/obs/`, allow `/collections/test-data/obs/`, allow `/collections/obs/` etc. <br /> 

   Rule is simply written as: <br />
   `allow { 
      input.request_path = /collections/...
   }` <br />  and the rule object is:
   ```json     
   {
      "command": "input_prop_equals",
      "properties": {
         "input_property": "request_path",
         "value": ["v1", "collections", "*"]
      }
   }

>    ```

Reviewer: Rephrase.

Example:

This logic handles all the paths after a particular section. Let's look at an example rule:

---json file---

In the JSON above, the `value` key contains an asterik in the values section. This indicates that the endpoint x is allowed to do y.

---rego equivalent---

Same applies for every other command in this section.

 {
      "command": "input_prop_in_as",
      "properties": {
         "datasource_name": "items",
         "datasource_loop_variables": ["name", "everyone"],
         "input_properties": ["preferred_username", "groupname"]
      }
 }
   ```

## 4. allow_full_access
   This command is to allow full access to the resource, if the property on the request object has a particular value<br />

   `allow {
  input.groupname == "EDITOR_ATAC"
}
   }` <br /> 
   and the rule object is:
   ```json
   [
      {
         "command": "allow_full_access",
         "properties": {
            "input_property": "groupname",
            "value": "EDITOR_ATAC",
         },
      }
   ]
   ```

   API CRUD Operations
   ============

   The API is designed to support all CRUD operations that makes interaction with the policies easy. The following routes are currently defined:

 ###  - POST `/policies/`  🗝
Using the JSON defined from the frontend, this route is used to build a new policy. The response will be a REGO file written and pushed to github as a newly established remote repository, with the request body conforming to a specified syntax described by the pydantic model `Policy`. <br />

> ```py3

Reviewer: Instead:

The `/policies` route receives a JSON request body comprising the rules needed to create a new policy. The policy upon a successful operation is created and the defined GitHub repository is updated with the policy file of _.rego_ format.

Anyone can choose to use the API and not necessarily a frontend..

class Policy:
    name: str
    rules: List[List[Rule]]
```
Rule Object: <br />
```py3
class Rule(BaseModel):
    command: str
    properties: Dict[str, Union[str, List[str], Dict[str, Union[str, List[str]]]]]
```

Each Rule object forms a specific rule in a Allow block, and a list of Rules forms a Allow block. <br />
```py3
List[Rule] == Allow {
   ...
}

> ```

Reviewer: The definition of the rule object should be before the explanations of the path operations.

> A typical request body which will contain multiple allow blocks in the REGO rule would be: <br />

Reviewer: Also, can we make the example rule shorter?

         }
      ],
      [
         {
         "command": "allow_full_access",
         "properties": {
            "input_property": "groupname",
            "value": "EDITOR_ATAC"
         }
         }
      ],
      [
         {
         "command": "input_prop_equals",
         "properties": {
            "input_property": "groupname",
            "value": [
               "v1",
               "collections",
               "test-data"
            ]
         }
         },
         {
         "command": "allow_full_access",
         "properties": {
            "input_property": "name",
            "value": "admin"
         }
         }
      ],
      [
         {
         "command": "input_prop_equals",
         "properties": {
            "input_property": "request_path",
            "value": [
               "v1",
               "collections",
               "lakes" ]
            }
         }
      ]
   ]}

   ```
 ###  - GET `/policies/` 🗝
This route is used to get all policies that have been created. The response will be a list of all policies that have been created by a certain user, and contains all the associating rules with the policy <br />

```py3
List[Policy]   

>    ```

Reviewer: Not needed here.

   ###  - GET PUT DELETE `/policies/{policy_name}` 🗝
   GET - This request method is used to get a specific policy by name. The response will be a policy object conforming to the pydantic model `Policy`. <br />
   PUT - This request method is used to update a specific policy by name. The response will be a policy object conforming to the pydantic model `Policy`. <br />

>    DELETE - This request method is used to delete a specific policy by name

Reviewer: They all need to show:

- Sample requests and responses
- What happens when a policy isn't found?

```py3
List[Rule] == Allow {
   ...
}
```

The API supports the following commands; input_prop_equals, input_prop_in, input_prop_in_as, allow_full_access.

Usage
============
### Detailed explanation of commands with examples: 
This section contains detailed explanations of the commands and examples of how to use them. 
## 1. Input_props_equals
   This command has different logic to handle series of equality checks.
  ### - Handling '*' as the wildcard flag: <br /> 

   This logic handles all the paths after a particular section. if `/collections/` is supplied as the option, all the routes after it will be allowed e.g allow `/collections/obs/`, allow `/collections/test-data/obs/`, allow `/collections/obs/` etc. <br /> 

   Example:
   ```json     
   {
      "command": "input_prop_equals",
      "properties": {
         "input_property": "request_path",
         "value": ["v1", "collections", "*"]
      }
   }
   ```
   In the json above, the `value` key holds an asterik in the values section, to indicate that the endpoint x is allowed to do y.

   The REGO equivalent of the above rule object is: <br />

>    `input.request_path[0] == "v1" && input.request_path[1] == "collections"`

Reviewer: Put this in a code block.

   This logic handles cases where a particular property on the input object is to be checked for equality against a value. 
   Example:

   ```json
   {
      "command": "input_prop_equals",
      "properties": {
         "input_property": "company",
         "value": "Geobeyond srl"
      }
   }
   ```
   In the json above, the `value` key holds the value that is to be checked for equality, while the `input_property` key holds the property that is to be checked in the input object.

   The REGO equivalent of the above rule object is: <br />
   `input.company == "Geobeyond srl"`

 ### -  Allow access to a particular path: <br />
 This logic handles cases where a specific path is to granted access, if certain property is present on the input object.

   Example:
   ```json
   [
      {
         "command": "input_prop_equals",
         "properties": {
            "input_property": "request_path",
            "value": ["v1", "collections", "lakes"],
         },
      },
       {
         "command": "input_prop_equals",
         "properties": {
            "input_property": "request_path",
            "value": "admin",
         },
      }
   ]
   ```

> In the json above, the list of objects indicate a very speciall `allow` block, that combines two commands. The `input_property` key holds the property that is to be va;idated before the request to that path is passed. 

Reviewer: ```suggestion
In the JSON above, the list of objects indicates a very special `allow` block, that combines two commands. The `input_property` key holds the property that is to be validated before the request to that path is passed. 
```

The REGO equivalent of the above rule object is: <br />
   `allow {

>   input.request_path == ["v1", "collections", "lakes", ""] && input.groupname == "admin"

Reviewer: indent properly.

}` 

>    ## 2. Input_props_in

Reviewer: Remove number and it can't start with a block letter.

   This logic checks if a particular property on the input object is present in a list of values from the database. <br />

   Example:

   ```json
   {
      "command": "input_prop_in",
      "properties": {
         "input_property": "company",
         "datasource_name": "items",
         "datasource_loop_variable": "name"
      }
}
   ```

>    In the json above, the `input_property` key holds the property that is to be validated before the request to that path is passed. The `datasource_name` key holds the name of the datasource(a list) from the database. The `datasource_loop_variable` key holds the name of the key on each object in the datasource.

Reviewer: Might be good to give a sentence or two about what data source items are and how they're gotten - from a database or smtn.

   The REGO equivalent of the above rule object is: <br />
   `input.company == data.items[i].name`

>    ## 3. Input_props_in_as

Reviewer: Remove number and word shouldn't start with a block letter.

   This logic checks if the value of two properties on the input object is present on one object in the database <br />

   Example:

   ```json
 {
      "command": "input_prop_in_as",
      "properties": {
         "datasource_name": "items",
         "datasource_loop_variables": ["name", "everyone"],
         "input_properties": ["preferred_username", "groupname"]
      }
 }
   ```
   In the json above, the resulting REGO code loops over the datasource twice, checking for equality between the values of the input properties, and the values of the datasource loop variables.

   The REGO equivalent of the above rule object is: <br />
         `some i` <br />
         `data.items[i].name == input.preferred_username` <br />

>          `data.items[i].everyone == input.groupname` <br /> 

Reviewer: Place it in a codeblock.

> ## 4. allow_full_access

Reviewer: Remove number.

>    This logic allows full access to the resources defined. if the value of the property on the input object has a particular value<br />

Reviewer: ```suggestion
   This logic allows full access to the resources defined. If the value of the property on the input object has a particular value<br />
```

   Example:
   ```json
   [
      {
         "command": "allow_full_access",
         "properties": {
            "input_property": "groupname",
            "value": "EDITOR_ATAC",
         },
      }
   ]
   ```

>    In the json above, the result is an allow block, that allows access to the resources defined, if the value of the property on the input object has a particular value.

Reviewer: ```suggestion
   In the JSON above, the result is an allow block that allows access to the resources defined if the value of the property on the input object has a particular value.
```

   `allow { ` <br />
      `input.groupname == "EDITOR_ATAC"` <br />
   `}` <br />

   API CRUD Operations
   ============

   The API is designed to support all CRUD operations that makes interaction with the policies easy. The following routes are currently defined:

 ###  - POST `/policies/`  🗝

> Using the JSON defined from the frontend, this route is used to build a new policy. The response will be a REGO file written and pushed to github as a newly established remote repository, with the request body conforming to a specified syntax described by the pydantic model `Policy`. <br />

Reviewer: ```suggestion
Using the JSON defined from the frontend, this route is used to build a new policy. The response will be a REGO file written and pushed to GitHub as a newly established remote repository, with the request body conforming to a specified syntax described by the pydantic model `Policy`. <br />
```

            }
            },
            {
               "command": "input_prop_in",
               "properties": {
                  "input_property": "company",
                  "datasource_name": "items",
                  "datasource_loop_variable": "name"
               }
            },
            {
               "command": "input_prop_equals",
               "properties": {
                  "input_property": "request_method",
                  "value": "GET"
               }
            }
         ],
         [
               {
               "command": "allow_full_access",
               "properties": {
                  "input_property": "groupname",
                  "value": "EDITOR_ATAC"
               }
            }
         ]
      ]}

   ```

   An example response body is: <br />
   ```json
   {"status": 200, "message": "Policy created successfully"}
   ```
   ###  - GET `/policies/` 🗝
   This route is used to get all policies that have been created. The response will be a list of all policies that have been created by a certain user, and contains all the associating rules with the policy <br />

   An example response body is: <br />
   ```json
   [
      {

>          "name": "Example",

Reviewer: The name should match the one in the POST request sent above.

         "rules": [
            [
               {
               "command": "input_prop_equals",
               "properties": {
                  "input_property": "request_path",
                  "value": [
                     "v1",
                     "collections",
                     "*"
                  ],
                  "exceptional_value": "obs"
               }
               },
               {
                  "command": "input_prop_in",
                  "properties": {
                     "input_property": "company",
                     "datasource_name": "items",
                     "datasource_loop_variable": "name"
                  }
               },
               {
                  "command": "input_prop_equals",
                  "properties": {
                     "input_property": "request_method",
                     "value": "GET"
                  }
               }
            ],
            [
               {
               "command": "allow_full_access",
               "properties": {
                  "input_property": "groupname",
                  "value": "EDITOR_ATAC"
               }
            }
         ]
      }
   ]
   ```

>    ###  - GET PUT DELETE `/policies/{policy_name}` 🗝

Reviewer: Can you put each operation in a different section as opposed to this?

      3. [Input_props_in_as](#input_props_in_as)
      4. [allow_full_access](#allow_full_access)
   * [API CRUD operations](#api-crud-operations)

  Introduction
============
This API is responsible for converting JSON variables to REGO which in turn is used in application policy management. The policy management application is born from the need to manage Open Policy Agent ( OPA ) policies written in REGO from a user interface for non-technical users and an API through JSON for technical users.

  Features
============
   - Translating JSON to REGO
   - CRUD operations for policies for persistence.
   - Supports existing policies for editing.

   Architecture
   ============
   The architecture of the API is based on the following principles:
   - The API is written in Python and FastAPI framework.
   - The API is written in a way that it is easy to use, extend, test and maintain.

![Screenshot 2022-06-13 at 09 55 43](https://user-images.githubusercontent.com/67229938/173343113-d51d72b4-84c8-4c3b-8555-af41e59cd2de.png)

View the Architecture in the browser [Here](https://www.figma.com/file/684S7kO4dPQbbZFZr6xZOn/Rego-builder?node-id=0%3A1) 

   Installation

>    ============

Reviewer: We can add a line to tell users that there are a number of ways to install the application.

### - Manual installation 
- Create a virtual environment and install the application dependencies:

```console
$ poetry shell && poetry install
```

- Configure your environment variables:

```dotenv
BASE_PATH = /tmp/fastgeoapi
DATABASE_PATH = /tmp/fastgeoapi/database.json
GITHUB_ACCESS_TOKEN=`cat ~/.github_access_token`
GITHUB_USERNAME=<your_github_username>
GITHUB_URL=<your_github_url where the authorization code lives>
CLIENT_ID=<your_github_client_id>
CLIENT_SECRET=<your_github_client_secret>
SECRET_KEY=<your_secret_key>
ALGORITHM=<your_algorithm e.g HS256>
ENVIRONMENT=<your_environment e.g. production|development>
```

> - Run the application from the entry point

Reviewer: ```suggestion
- Run the application from the entry point:
```

```console
$ python3 main.py
```
- Open [localhost:8000/docs](localhost:8000/docs) for API Documentation

### - Build the application image and start the container -
   ```console
   $ Configure environment variables - the .env file
   $ docker-compose -f docker-compose.dev.yml up -d

   ```
### - Pull the official image from Docker Hub and start the container 

   ```console
   $ docker pull rscheele3214/rego-builder:latest
   $ docker run rscheele3214/rego-builder:latest

>    ```

Reviewer: Why are there indentations in the document?

   ```json     
   {
      "command": "input_prop_equals",
      "properties": {
         "input_property": "request_method",
         "value": "GET"
      }
   }
   ```
In the example above, the command key represents the operation to be performed and the properties key represents the properties that are being used in the operation.

`input_prop_equals` is the command in the example that initiates the appropriate operation, on the properties object.
The above rule translates to an equality check between the input property `input_property` and the value `value`. <br /> <br />
The REGO equivalent of the above rule object is: <br />
```rego
   `input.request_method == "GET"`
```

Each Rule object forms a specific rule in a Allow block, and a list of Rules forms a Allow block. <br />
```py3
List[Rule] == Allow {
   ...
}
```

The API supports the following commands; input_prop_equals, input_prop_in, input_prop_in_as, allow_full_access.

Usage
============

> ### Detailed explanation of commands with examples: 

Reviewer: I think we can add a sentence about what this section entails instead of presenting another heading.

This section contains detailed explanations of the commands and examples of how to use them. 
### input_props_equals
   This command has different logic to handle series of equality checks.
  - Handling '*' as the wildcard flag: <br /> 

>    This logic handles all the paths after a particular section. if `/collections/` is supplied as the option, all the routes after it will be allowed e.g allow `/collections/obs/`, allow `/collections/test-data/obs/`, allow `/collections/obs/` etc. <br /> 

Reviewer: Remove indents.

   Example:
   ```json     
   {
      "command": "input_prop_equals",
      "properties": {
         "input_property": "request_path",
         "value": ["v1", "collections", "*"]
      }
   }

>    ```

Reviewer: Remove indents.

   "properties": {
      "input_property": "request_path",
      "value": ["v1", "collections", "*"],
      "exceptional_value": "obs",
   },
}
```
In the json above, the `value` key holds an asterik in the values section, to indicate that the endpoint x is allowed to do y. The `exceptional_value` key holds the value of the path parameter that is to be exempted.

The REGO equivalent of the above rule object is: <br />
```rego
input.request_path[0] == "v1"
input.request_path[1] == "collections"
input.request_path[2] != "obs"
```

-  Handling equality check between a particular property on the request object and a value: <br />
This logic handles cases where a particular property on the input object is to be checked for equality against a value. 
Example:

```json
{
   "command": "input_prop_equals",
   "properties": {
      "input_property": "company",
      "value": "Geobeyond srl"
}
}
```
In the json above, the `value` key holds the value that is to be checked for equality, while the `input_property` key holds the property that is to be checked in the input object.

The REGO equivalent of the above rule object is: <br />
```rego
input.company == "Geobeyond srl"
```

> 
> 

Reviewer: ```suggestion
```

-  Allow access to a particular path: <br />
This logic handles cases where a specific path is to granted access, if certain property is present on the input object.

Example:
```json
[
{
   "command": "input_prop_equals",
   "properties": {
      "input_property": "request_path",
      "value": ["v1", "collections", "lakes"],
   },
},
   {
   "command": "input_prop_equals",
   "properties": {
      "input_property": "request_path",
      "value": "admin",
   },
}
]
```
In the JSON above, the list of objects indicates a very special `allow` block, that combines two commands. The `input_property` key holds the property that is to be validated before the request to that path is passed. 

The REGO equivalent of the above rule object is: <br />
```rego
allow {
   input.request_path == ["v1", "collections", "lakes", ""]
   input.groupname == "admin"
}

> ```

Reviewer: ```suggestion
```

```

### input_props_in
This logic checks if a particular property on the input object is present in a list of values from the database. <br />

Example:

```json
{
"command": "input_prop_in",
"properties": {
   "input_property": "company",
   "datasource_name": "items",
   "datasource_loop_variable": "name"
}
}
```
In the json above, the `input_property` key holds the property that is to be validated before the request to that path is passed. The `datasource_name` key holds the name of the datasource(a list) from the database. The `datasource_loop_variable` key holds the name of the key on each object in the datasource.

The rego rules combine data from the database with the input object, to work out certain conditions. The datasource is the list of values from the database, and the `datasource_loop_variable` is the name of the key on each object in the datasource.

The REGO equivalent of the above rule object is: <br />
```rego
input.company == data.items[i].name
```

> ### input_props_in_as

Reviewer: ```suggestion

### input_props_in_as
```

>    The rego rules combine data from the database with the input object, to work out certain conditions. The datasource is the list of values from the database, and the datasource_loop_variable is the name of the key on each object in the datasource.

Reviewer: ```suggestion
   The rego rules combine data from the database with the input object, to work out certain conditions. The datasource is the list of values from the database, and the `datasource_loop_variable` is the name of the key on each object in the datasource.
```

This logic checks if the value of two properties on the input object is present on one object in the database <br />

Example:
```json
{
   "command": "input_prop_in_as",
   "properties": {
      "datasource_name": "items",
      "datasource_loop_variables": ["name", "everyone"],
      "input_properties": ["preferred_username", "groupname"]
   }
}
```

> In the json above, the resulting REGO code loops over the datasource twice, checking for equality between the values of the input properties, and the values of the datasource loop variables.

Reviewer: ```suggestion

In the JSON above, the resulting REGO code loops over the datasource twice, checking for equality between the values of the input properties, and the values of the datasource loop variables.
```

>    ```

Reviewer: Format this section properly.

The REGO equivalent of the above rule object is: <br />
```rego
some i
data.items[i].name == input.preferred_username
data.items[i].everyone == input.groupname
```

### allow_full_access
This logic allows full access to the resources defined. If the value of the property on the input object has a particular value<br />

Example:
```json
[
   {
      "command": "allow_full_access",
      "properties": {
         "input_property": "groupname",
         "value": "EDITOR_ATAC",
      },
   }
]
```
In the JSON above, the result is an allow block that allows access to the resources defined if the value of the property on the input object has a particular value.

```rego
allow {
   input.groupname == "EDITOR_ATAC"
}
```

> 
> 
> 

Reviewer: ```suggestion
```

POST `/policies/`  Create new policy
============

> Using the JSON defined from the frontend, this route is used to build a new policy. The response will be a REGO file written and pushed to GitHub as a newly established remote repository, with the request body conforming to a specified syntax described by the pydantic model `Policy`. <br />

Reviewer: Again, we should not make it seem like only the frontend can send JSON. A better way to start this section:

The POST route takes a request body containing the rules defined according to the schema.
